fix(users): GitHub strategy missing email

mleanos opened the pull request:
Fixes an issue with an empty/missing/null Email coming from GitHub's
OAuth call response.

Also, introduces the `sparse` index option on the User model's Email
field. This will ensure that we can have multiple User documents without
the Email field.

Related #1145

mleanos commented:
I'll add some tests to this PR, sometime tomorrow.

mleanos commented:
Added a test to ensure the `sparse` option is working properly.

@lirantal WDYT?

lirantal commented:
I think it's overall ok, although I'm wondering if there's really a case where GitHub will return no email for a user?

mleanos commented:
If a user chooses the "Don't show my email address" option in their Profile settings for Public email. Then GitHub won't return the email address. Also, Twitter doesn't send user's email at all.

lirantal commented:
Which only make you wonder how twitter ever worked, right? :)
Ok, let's see if anyone else has anything else to contribute

mleanos commented:
Twitter worked, because we allow empty/missing email fields in our User model. The issue will only show itself if there's an attempt to save a new User without the email field. Thus, mongodb treats both of those as the same, and won't allow it since we have the unique index setting on that field; there's a violation of the unique constraint in that case.

We don't run into this issue with local testing, since most of us just use one User for any particular Social Account. I guess User's of this framework haven't ran into this issue, or reported it. Until @farajfarook did.

lirantal commented:
That's what I was referring to - that probably users of MEAN.JS who used GitHub did get this problem for such cases where the user opted out from sharing the email detail.

Anyway, it looks like a correct way of handling it.
LGTM.

@codydaig maybe another pair of eyes?

lirantal commented:
@mleanos BTW you are definitely able to run some manual tests with local users signup, GitHub as well as another strategy, right? just to make sure we're not breaking anything larger here.

mleanos commented:
@lirantal I've been going through the Social Accounts implementation & testing. I've noticed some issues with the other providers as well. This fix is part of a bigger initiative to get the Social Providers working 100%. I have a list of issues I've come across, but haven't organized it yet.

I have tested this with the other provider's and there aren't any issues, other than the other aspects of the Social Provider's implementation that are broken. However, this does also take care of the issue raised in #1145. 

I did extensive testing with the `sparse` option, and it won't have any adverse effects. The only nuance is that when this is integrated into an existing application, the index on the `email` field in the MongoDB instance would need to be drooped. The next time the application is run, Mongoose will rebuild the index using the `sparse` option. I added a comment in the Schema for that.

Also, the test I added will ensure that the `sparse` index works; regardless of which Provider is being used.

lirantal commented:
@mleanos ok, that's clear.
Do users have to manually drop the index or will mongodb rebuild it automatically for them?
If it's manually, I'm not sure users will understand that straight away - is it possible to catch the error from mongodb on that and show a message to guide them?

mleanos commented:
User's do have to manually drop the index. However, we can probably add a Gulp task that is executed with one of our other tasks, that start the server; `default`, `prod`, `test`. Do you think that would work? 

Another option might be to add something to our mongoose server config, that ensures the index is dropped if the sparse option is not set for it. Or we simply display a message in this case. I haven't looked into that, but I think it will be possible. One caveat to this option is that this check will be performed every time the database is started; I'm not sure how I feel about that.

I'd like to make it as easy as possible for our users. However, it might be more trouble for us than it's worth. As it stands, it's technically not a breaking change. Users will only run into any issue when a duplicate `user` is saved that violates the indexes unique constraint; in this case, a missing/empty `email` field.

I'm happy to do more work on this to make it easier on our users. But I just wanted to throw those options out there. And the inline comment might suffice for now.

lirantal commented:
We would like to be unobtrusive but in the same time not break things horribly.
This has been a recurrent issue for us with other updates we had before like fixing the `users` table's salt which was accidentally binary and that messed things up and was a breaking change too when we fixed it.

I would like to suggest that we have an `UPGRADE.md` (and possibly accompanying wiki page or whatever else helps as well) that states the big changes that require user involvement when the version changes.

@ilanbiala WDYT?

ilanbiala commented:
If we do like Angular does and put in our Changelog the breaking changes and the fixes, then I think that is sufficient.

lirantal commented:
sounds good to me.
@mleanos can you include this kind of update with this PR about the index change and instructions of what should be done?

mleanos commented:
I added a `UPGRADE.md` to this PR. However, I may need help with the markup. I don't have a way of testing the markup locally, since I'm on Windows. And AFAIK Jenkins isn't supported on Windows.

lirantal commented on UPGRADE.md:
```diff
@@ -0,0 +1,12 @@
+  * If you are upgrading an existing MEANJS application, and have not changed this field, you will already have an existing index on the Email field of your Users collection. If this is the case, you will need to follow these steps for the `sparse` option to be applied.
+        
+
+    1. Manually drop the Email field's index in your Users collection, on your applications MongoDB instance.
```
@mleanos 
Can we provide actual copy&paste instructions on how to do it to ease the users?

mleanos replied:
Sure. I can add the MongoDB commands to drop the index, and then the commands to start the application.

lirantal commented:
@mleanos added my comment and also you can see how the markup renders through GitHub, check this URL for your file: https://github.com/mleanos/mean/blob/fix/users-github-strategy-email/UPGRADE.md

mleanos commented:
@lirantal I know how to test it once it's in a remote GitHub repo. I just can't view the rendered markup locally. It's not a problem. I can just push to my fork, and test there.

mleanos commented:
@lirantal @ilanbiala @codydaig I updated this PR to include a companion upgrade script, for dropping the index using Mongoose. I also updated the UPGRADE readme to include directions on how to use the script.

How does the UPGRADE.md file look to y'all?

codydaig commented on scripts/upgrade-users-sparse-index.js:
```diff
@@ -0,0 +1,59 @@
+
+var mongoose = require('mongoose'),
+  chalk = require('chalk'),
+  mg = require('../config/lib/mongoose');
```
I think this should be mongoose instead of mg to be consistent with the rest of the project.

ilanbiala replied:
mongoose is defined above

simison replied:
`mongooseConfig`, then?

codydaig commented:
Upgrade.md looks fine to me. In the upgrade script, I don't want to start abbreviating package names. I'd rather keep it as `mongoose`.

mleanos commented:
@codydaig I addressed your comment. For clarification, `mg` was referencing our mongoose configuration & `mongoose` was already used by the reference to the Mongoose library; the latter was not used in this script so I removed it and renamed the other reference :)

@lirantal How does this look to you now, in regards to your comments about the upgrade instructions?

codydaig commented:
LGTM

lirantal commented:
Great, let's add.

ilanbiala commented on scripts/upgrade-users-sparse-index.js:
```diff
@@ -0,0 +1,58 @@
+  // get a reference to the User collection
+  var userCollection = db.connections[0].collections.users;
+
+  console.log(chalk.yellow('*************** Removing index "' + _indexToRemove + '" from the User collection ***************'));
```
Is this line not over 80 characters?

mleanos replied:
@ilanbiala Yes. I don't really like what I did with the messages. I'll clean them up.

Just to clarify, are you concerned with the length of the line of the code, or the output of the console log?

ilanbiala replied:
Mostly length, I think the content is pretty good.

mleanos replied:
Alright. I can break these up into multiple lines of code.

ilanbiala commented on scripts/upgrade-users-sparse-index.js:
```diff
@@ -0,0 +1,58 @@
+      message = 'An error occured while removing the index "' + _indexToRemove + '".';
+
+      if (err.message.indexOf('index not found with name') !== -1) {
+        message = 'An index with the name "' + _indexToRemove + '" could not be found. Please double check the index name in your mongodb User collection.';
```
over 80 characters?

ilanbiala commented on scripts/upgrade-users-sparse-index.js:
```diff
@@ -0,0 +1,58 @@
+    }
+  } else {
+    console.log(chalk.green(message));
+    console.log(chalk.green('The next time your application starts, Mongoose will perform ensureIndex() which will rebuild the index "' + _indexToRemove + '".'));
```
over 80 characters?

mleanos commented:
@ilanbiala I've cleaned up those long lines of code. LGTY?

ilanbiala commented on scripts/upgrade-users-sparse-index.js:
```diff
@@ -0,0 +1,60 @@
+    }
+  } else {
+    console.log(chalk.green(message));
+    console.log(chalk.green('The next time your application starts, Mongoose will rebuild the index "' + _indexToRemove + '".'));
```
This line is too long still.

mleanos replied:
Forgot that one :|

trendzetter commented:
:+1:

ilanbiala commented on modules/users/tests/server/user.server.model.tests.js:
```diff
@@ -193,6 +193,29 @@ describe('User Model Unit Tests:', function () {
 
     });
 
+    it('should not index missing email field, and not enforce unique index constraint (sparse index)', function (done) {
```
Also slightly long, this is 120 chars

ilanbiala commented on scripts/upgrade-users-sparse-index.js:
```diff
@@ -0,0 +1,60 @@
+  var userCollection = db.connections[0].collections.users;
+
+  console.log();
+  console.log(chalk.yellow('Removing index "' + _indexToRemove + '" from the User collection.'));
```
97 chars

ilanbiala commented:
@mleanos strangely we don't have a linting rule for line length. Adding that would be nice for the future so we can catch extra long lines more easily.

mleanos commented:
@ilanbiala I shortened the lines in the upgrade script to be less than 80 chars. As for the test line you mentioned, I shortened it as much as possible but it's still about 105 chars long. I didn't see any other way of shortening that line. Quite a bit of our `describe()` lines are longer than 80 chars.

If this looks good to you now, I'll merge.

ilanbiala commented:
LGTm.
